src/main/java/net/snowflake/ingest/connection/OAuthClient.java

```diff
@@ -4,11 +4,122 @@
+import net.snowflake.client.jdbc.internal.apache.http.entity.ContentType;
+import net.snowflake.client.jdbc.internal.apache.http.entity.StringEntity;
+import net.snowflake.client.jdbc.internal.apache.http.impl.client.CloseableHttpClient;
+import net.snowflake.client.jdbc.internal.apache.http.util.EntityUtils;
+import net.snowflake.client.jdbc.internal.google.api.client.http.HttpStatusCodes;
+import net.snowflake.client.jdbc.internal.google.gson.JsonObject;
+import net.snowflake.client.jdbc.internal.google.gson.JsonParser;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.HttpUtil;
+import net.snowflake.ingest.utils.SFException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-/** Interface to perform token refresh request from {@link OAuthManager} */
-public interface OAuthClient {
-  AtomicReference<OAuthCredential> getoAuthCredentialRef();
+/*
+ * Implementation of OAuth Client, used for refreshing an OAuth access token.
+ */
+public class OAuthClient {
+
+  static final Logger LOGGER = LoggerFactory.getLogger(OAuthClient.class);
+  private static final String TOKEN_REQUEST_ENDPOINT = "/oauth/token-request";
+
+  // Content type header to specify the encoding
+  private static final String OAUTH_CONTENT_TYPE_HEADER = "application/x-www-form-urlencoded";
+  private static final String GRANT_TYPE_PARAM = "grant_type";
+  private static final String ACCESS_TOKEN = "access_token";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String EXPIRES_IN = "expires_in";
+
+  // OAuth credential
+  private final AtomicReference<OAuthCredential> oAuthCredential;
+
+  // Http client for submitting token refresh request
+  private final CloseableHttpClient httpClient;
+
+  /**
+   * Creates an AuthClient for Snowflake OAuth given account, credential and base uri
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param baseURIBuilder - the uri builder with common scheme, host and port
+   */
+  OAuthClient(String accountName, OAuthCredential oAuthCredential, URIBuilder baseURIBuilder) {
+    this.oAuthCredential = new AtomicReference<>(oAuthCredential);
+
+    // build token request uri
+    baseURIBuilder.setPath(TOKEN_REQUEST_ENDPOINT);
+    this.httpClient = HttpUtil.getHttpClient(accountName);
+  }
+
+  /** Get access token */
+  public AtomicReference<OAuthCredential> getOAuthCredentialRef() {
+    return oAuthCredential;
+  }
+
+  /** Refresh access token using a valid refresh token */
+  public void refreshToken() {
+    String respBodyString = null;
+    try (CloseableHttpResponse httpResponse = httpClient.execute(makeRefreshTokenRequest())) {
+      respBodyString = EntityUtils.toString(httpResponse.getEntity());
+
+      if (httpResponse.getStatusLine().getStatusCode() == HttpStatusCodes.STATUS_CODE_OK) {
```

Comment:
What happens if the status code is not `STATUS_CODE_OK`?

Reply:
Fixed the exception flow.